src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs

```diff
@@ -121,6 +180,29 @@ public DiagnosticAnalysisResult ToAggregatedForm()
             return new DiagnosticAnalysisResult(ProjectId, Version, DocumentIds, IsEmpty, FromBuild);
         }
 
+        public DiagnosticAnalysisResult UpdateAggregatedResult(VersionStamp version, DocumentId documentId, bool fromBuild)
+        {
+            return new DiagnosticAnalysisResult(ProjectId, version, DocumentIdsOrEmpty.Add(documentId), isEmpty: false, fromBuild: fromBuild);
+        }
+
+        public DiagnosticAnalysisResult Reset()
+        {
+            return new DiagnosticAnalysisResult(ProjectId, VersionStamp.Default, DocumentIds, IsEmpty, FromBuild);
```

Comment:
this is subtle to me.  the fact that 'FromBuild' and 'IsEmpty' are preserved are not necessairly things i would expect.

Reply:
also, it's not clear to me if 'IsEmpty' is actually correct here.  IsEmpty above is computed as ` IsEmpty = DocumentIds.IsEmpty && _others.IsEmpty;`.  But we don't pass '_others' here.  So we may have been non-empty before, but we should be empty now.  Soemthing seems smelly here.